File: src/main/java/pramp/GetDifferentNumber.java
package main.java.pramp;

import java.util.HashSet;
import java.util.Set;

public class GetDifferentNumber {
    static int getDifferentNumber(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int v = arr[i];
            // if within bounds and v is not already in its supposed place
            if (v < arr.length && v != i) {
                // swap these
                int temp = arr[v];
                arr[v] = arr[i];
                arr[i] = temp;
            }
        }

        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != i) return i;
        }

        // return the size since arr is ordered starting from 0
        return arr.length;
    }

    static int getDifferentNumber2(int[] arr) {
        // your code goes here
        Set<Integer> set = new HashSet<>();
        int minVal = Integer.MAX_VALUE;
        for (int value : arr) {
            set.add(value);
            minVal = Math.min(minVal, value);
        }

        // For cases where min value is bigger than 0
        if (minVal > 0) return 0;

        int result = 0;
        while (set.contains(result)) result++;

        return result;
    }

    static int getDifferentNumberImproved(int[] arr) {
        int i = 0;
        while (i < arr.length) {
            if (arr[i] < arr.length && arr[i] != i && arr[arr[i]] != arr[i]) {
                // Swap arr[i] and arr[arr[i]]
                int temp = arr[arr[i]];
                arr[arr[i]] = arr[i];
                arr[i] = temp;
            } else {
                // Move to the next index only if the current number is in its correct place
                // or cannot be placed.
                i++;
            }
        }

        // After rearrangement, the first index that does not match its value is the answer.
        for (i = 0; i < arr.length; i++) {
            if (arr[i] != i) {
                return i;
            }
        }

        // If all indices are correct, the answer is arr.length.
        return arr.length;
    }

    public static void main(String[] args) {
        System.out.println(getDifferentNumber2(new int[]{0}));
        System.out.println(getDifferentNumber2(new int[]{0, 1, 2}));
        System.out.println(getDifferentNumber2(new int[]{1, 3, 0, 2}));
        System.out.println(getDifferentNumber2(new int[]{0, 5, 4, 1, 3, 6, 2}));
        System.out.println(getDifferentNumber2(new int[]{1000000}));
        System.out.println();

        System.out.println(getDifferentNumber(new int[]{0}));
        System.out.println(getDifferentNumber(new int[]{0, 1, 2}));
        System.out.println(getDifferentNumber(new int[]{1, 3, 0, 2}));
        System.out.println(getDifferentNumber(new int[]{0, 5, 4, 1, 3, 6, 2}));
        System.out.println(getDifferentNumber(new int[]{1000000}));
        System.out.println();

        System.out.println(getDifferentNumberImproved(new int[]{0}));
        System.out.println(getDifferentNumberImproved(new int[]{0, 1, 2}));
        System.out.println(getDifferentNumberImproved(new int[]{1, 3, 0, 2}));
        System.out.println(getDifferentNumberImproved(new int[]{0, 5, 4, 1, 3, 6, 2}));
        System.out.println(getDifferentNumberImproved(new int[]{1000000}));

    }
}

File: src/pramp/GetDifferentNumber.kt
package pramp


fun getDifferentNumber(arr: IntArray): Int {
    arr.forEachIndexed { i, v ->
        if (v < arr.size) {
            val temp = arr[v]
            arr[v] = v
            arr[i] = temp
        }
        while (arr[i] < i) {
            val temp = arr[arr[i]]
            arr[arr[i]] = arr[i]
            arr[i] = temp
        }
    }

    arr.forEachIndexed { i, v ->
        if (v != i) return i
    }
    // Every value is in its correct place
    return arr.size
}
fun getDifferentNumber1(arr: IntArray): Int {
    val min = arr.min()
    if (min > 0) return 0

    val set = arr.toSet()
    for (i in arr.indices) {
        if (!set.contains(i)) return i
    }

    return arr.size
}

fun main() {
    println(getDifferentNumber(intArrayOf(0)))
    println(getDifferentNumber(intArrayOf(0, 1, 2, 3)))
    println(getDifferentNumber(intArrayOf(0, 5, 4, 1, 3, 6, 2)))
    println(getDifferentNumber(intArrayOf(0, 100000)))
    println(getDifferentNumber(intArrayOf(1000000)))
    println("----")
    println(getDifferentNumber1(intArrayOf(0)))
    println(getDifferentNumber1(intArrayOf(0, 1, 2, 3)))
    println(getDifferentNumber1(intArrayOf(0, 5, 4, 1, 3, 6, 2)))
    println(getDifferentNumber1(intArrayOf(0, 100000)))
    println(getDifferentNumber1(intArrayOf(1000000)))
}
